<!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <title>load libraries</title>
        <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

/* From extension vscode.markdown-math */
@font-face{font-family:KaTeX_AMS;font-style:normal;font-weight:400;src:url(fonts/KaTeX_AMS-Regular.woff2) format("woff2"),url(fonts/KaTeX_AMS-Regular.woff) format("woff"),url(fonts/KaTeX_AMS-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Caligraphic;font-style:normal;font-weight:700;src:url(fonts/KaTeX_Caligraphic-Bold.woff2) format("woff2"),url(fonts/KaTeX_Caligraphic-Bold.woff) format("woff"),url(fonts/KaTeX_Caligraphic-Bold.ttf) format("truetype")}@font-face{font-family:KaTeX_Caligraphic;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Caligraphic-Regular.woff2) format("woff2"),url(fonts/KaTeX_Caligraphic-Regular.woff) format("woff"),url(fonts/KaTeX_Caligraphic-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Fraktur;font-style:normal;font-weight:700;src:url(fonts/KaTeX_Fraktur-Bold.woff2) format("woff2"),url(fonts/KaTeX_Fraktur-Bold.woff) format("woff"),url(fonts/KaTeX_Fraktur-Bold.ttf) format("truetype")}@font-face{font-family:KaTeX_Fraktur;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Fraktur-Regular.woff2) format("woff2"),url(fonts/KaTeX_Fraktur-Regular.woff) format("woff"),url(fonts/KaTeX_Fraktur-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:normal;font-weight:700;src:url(fonts/KaTeX_Main-Bold.woff2) format("woff2"),url(fonts/KaTeX_Main-Bold.woff) format("woff"),url(fonts/KaTeX_Main-Bold.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:italic;font-weight:700;src:url(fonts/KaTeX_Main-BoldItalic.woff2) format("woff2"),url(fonts/KaTeX_Main-BoldItalic.woff) format("woff"),url(fonts/KaTeX_Main-BoldItalic.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:italic;font-weight:400;src:url(fonts/KaTeX_Main-Italic.woff2) format("woff2"),url(fonts/KaTeX_Main-Italic.woff) format("woff"),url(fonts/KaTeX_Main-Italic.ttf) format("truetype")}@font-face{font-family:KaTeX_Main;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Main-Regular.woff2) format("woff2"),url(fonts/KaTeX_Main-Regular.woff) format("woff"),url(fonts/KaTeX_Main-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Math;font-style:italic;font-weight:700;src:url(fonts/KaTeX_Math-BoldItalic.woff2) format("woff2"),url(fonts/KaTeX_Math-BoldItalic.woff) format("woff"),url(fonts/KaTeX_Math-BoldItalic.ttf) format("truetype")}@font-face{font-family:KaTeX_Math;font-style:italic;font-weight:400;src:url(fonts/KaTeX_Math-Italic.woff2) format("woff2"),url(fonts/KaTeX_Math-Italic.woff) format("woff"),url(fonts/KaTeX_Math-Italic.ttf) format("truetype")}@font-face{font-family:"KaTeX_SansSerif";font-style:normal;font-weight:700;src:url(fonts/KaTeX_SansSerif-Bold.woff2) format("woff2"),url(fonts/KaTeX_SansSerif-Bold.woff) format("woff"),url(fonts/KaTeX_SansSerif-Bold.ttf) format("truetype")}@font-face{font-family:"KaTeX_SansSerif";font-style:italic;font-weight:400;src:url(fonts/KaTeX_SansSerif-Italic.woff2) format("woff2"),url(fonts/KaTeX_SansSerif-Italic.woff) format("woff"),url(fonts/KaTeX_SansSerif-Italic.ttf) format("truetype")}@font-face{font-family:"KaTeX_SansSerif";font-style:normal;font-weight:400;src:url(fonts/KaTeX_SansSerif-Regular.woff2) format("woff2"),url(fonts/KaTeX_SansSerif-Regular.woff) format("woff"),url(fonts/KaTeX_SansSerif-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Script;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Script-Regular.woff2) format("woff2"),url(fonts/KaTeX_Script-Regular.woff) format("woff"),url(fonts/KaTeX_Script-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size1;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size1-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size1-Regular.woff) format("woff"),url(fonts/KaTeX_Size1-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size2;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size2-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size2-Regular.woff) format("woff"),url(fonts/KaTeX_Size2-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size3;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size3-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size3-Regular.woff) format("woff"),url(fonts/KaTeX_Size3-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Size4;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Size4-Regular.woff2) format("woff2"),url(fonts/KaTeX_Size4-Regular.woff) format("woff"),url(fonts/KaTeX_Size4-Regular.ttf) format("truetype")}@font-face{font-family:KaTeX_Typewriter;font-style:normal;font-weight:400;src:url(fonts/KaTeX_Typewriter-Regular.woff2) format("woff2"),url(fonts/KaTeX_Typewriter-Regular.woff) format("woff"),url(fonts/KaTeX_Typewriter-Regular.ttf) format("truetype")}.katex{text-rendering:auto;font:normal 1.21em KaTeX_Main,Times New Roman,serif;line-height:1.2;text-indent:0}.katex *{-ms-high-contrast-adjust:none!important;border-color:currentColor}.katex .katex-version:after{content:"0.13.24"}.katex .katex-mathml{clip:rect(1px,1px,1px,1px);border:0;height:1px;overflow:hidden;padding:0;position:absolute;width:1px}.katex .katex-html>.newline{display:block}.katex .base{position:relative;white-space:nowrap;width:-webkit-min-content;width:-moz-min-content;width:min-content}.katex .base,.katex .strut{display:inline-block}.katex .textbf{font-weight:700}.katex .textit{font-style:italic}.katex .textrm{font-family:KaTeX_Main}.katex .textsf{font-family:KaTeX_SansSerif}.katex .texttt{font-family:KaTeX_Typewriter}.katex .mathnormal{font-family:KaTeX_Math;font-style:italic}.katex .mathit{font-family:KaTeX_Main;font-style:italic}.katex .mathrm{font-style:normal}.katex .mathbf{font-family:KaTeX_Main;font-weight:700}.katex .boldsymbol{font-family:KaTeX_Math;font-style:italic;font-weight:700}.katex .amsrm,.katex .mathbb,.katex .textbb{font-family:KaTeX_AMS}.katex .mathcal{font-family:KaTeX_Caligraphic}.katex .mathfrak,.katex .textfrak{font-family:KaTeX_Fraktur}.katex .mathtt{font-family:KaTeX_Typewriter}.katex .mathscr,.katex .textscr{font-family:KaTeX_Script}.katex .mathsf,.katex .textsf{font-family:KaTeX_SansSerif}.katex .mathboldsf,.katex .textboldsf{font-family:KaTeX_SansSerif;font-weight:700}.katex .mathitsf,.katex .textitsf{font-family:KaTeX_SansSerif;font-style:italic}.katex .mainrm{font-family:KaTeX_Main;font-style:normal}.katex .vlist-t{border-collapse:collapse;display:inline-table;table-layout:fixed}.katex .vlist-r{display:table-row}.katex .vlist{display:table-cell;position:relative;vertical-align:bottom}.katex .vlist>span{display:block;height:0;position:relative}.katex .vlist>span>span{display:inline-block}.katex .vlist>span>.pstrut{overflow:hidden;width:0}.katex .vlist-t2{margin-right:-2px}.katex .vlist-s{display:table-cell;font-size:1px;min-width:2px;vertical-align:bottom;width:2px}.katex .vbox{align-items:baseline;display:inline-flex;flex-direction:column}.katex .hbox{width:100%}.katex .hbox,.katex .thinbox{display:inline-flex;flex-direction:row}.katex .thinbox{max-width:0;width:0}.katex .msupsub{text-align:left}.katex .mfrac>span>span{text-align:center}.katex .mfrac .frac-line{border-bottom-style:solid;display:inline-block;width:100%}.katex .hdashline,.katex .hline,.katex .mfrac .frac-line,.katex .overline .overline-line,.katex .rule,.katex .underline .underline-line{min-height:1px}.katex .mspace{display:inline-block}.katex .clap,.katex .llap,.katex .rlap{position:relative;width:0}.katex .clap>.inner,.katex .llap>.inner,.katex .rlap>.inner{position:absolute}.katex .clap>.fix,.katex .llap>.fix,.katex .rlap>.fix{display:inline-block}.katex .llap>.inner{right:0}.katex .clap>.inner,.katex .rlap>.inner{left:0}.katex .clap>.inner>span{margin-left:-50%;margin-right:50%}.katex .rule{border:0 solid;display:inline-block;position:relative}.katex .hline,.katex .overline .overline-line,.katex .underline .underline-line{border-bottom-style:solid;display:inline-block;width:100%}.katex .hdashline{border-bottom-style:dashed;display:inline-block;width:100%}.katex .sqrt>.root{margin-left:.27777778em;margin-right:-.55555556em}.katex .fontsize-ensurer.reset-size1.size1,.katex .sizing.reset-size1.size1{font-size:1em}.katex .fontsize-ensurer.reset-size1.size2,.katex .sizing.reset-size1.size2{font-size:1.2em}.katex .fontsize-ensurer.reset-size1.size3,.katex .sizing.reset-size1.size3{font-size:1.4em}.katex .fontsize-ensurer.reset-size1.size4,.katex .sizing.reset-size1.size4{font-size:1.6em}.katex .fontsize-ensurer.reset-size1.size5,.katex .sizing.reset-size1.size5{font-size:1.8em}.katex .fontsize-ensurer.reset-size1.size6,.katex .sizing.reset-size1.size6{font-size:2em}.katex .fontsize-ensurer.reset-size1.size7,.katex .sizing.reset-size1.size7{font-size:2.4em}.katex .fontsize-ensurer.reset-size1.size8,.katex .sizing.reset-size1.size8{font-size:2.88em}.katex .fontsize-ensurer.reset-size1.size9,.katex .sizing.reset-size1.size9{font-size:3.456em}.katex .fontsize-ensurer.reset-size1.size10,.katex .sizing.reset-size1.size10{font-size:4.148em}.katex .fontsize-ensurer.reset-size1.size11,.katex .sizing.reset-size1.size11{font-size:4.976em}.katex .fontsize-ensurer.reset-size2.size1,.katex .sizing.reset-size2.size1{font-size:.83333333em}.katex .fontsize-ensurer.reset-size2.size2,.katex .sizing.reset-size2.size2{font-size:1em}.katex .fontsize-ensurer.reset-size2.size3,.katex .sizing.reset-size2.size3{font-size:1.16666667em}.katex .fontsize-ensurer.reset-size2.size4,.katex .sizing.reset-size2.size4{font-size:1.33333333em}.katex .fontsize-ensurer.reset-size2.size5,.katex .sizing.reset-size2.size5{font-size:1.5em}.katex .fontsize-ensurer.reset-size2.size6,.katex .sizing.reset-size2.size6{font-size:1.66666667em}.katex .fontsize-ensurer.reset-size2.size7,.katex .sizing.reset-size2.size7{font-size:2em}.katex .fontsize-ensurer.reset-size2.size8,.katex .sizing.reset-size2.size8{font-size:2.4em}.katex .fontsize-ensurer.reset-size2.size9,.katex .sizing.reset-size2.size9{font-size:2.88em}.katex .fontsize-ensurer.reset-size2.size10,.katex .sizing.reset-size2.size10{font-size:3.45666667em}.katex .fontsize-ensurer.reset-size2.size11,.katex .sizing.reset-size2.size11{font-size:4.14666667em}.katex .fontsize-ensurer.reset-size3.size1,.katex .sizing.reset-size3.size1{font-size:.71428571em}.katex .fontsize-ensurer.reset-size3.size2,.katex .sizing.reset-size3.size2{font-size:.85714286em}.katex .fontsize-ensurer.reset-size3.size3,.katex .sizing.reset-size3.size3{font-size:1em}.katex .fontsize-ensurer.reset-size3.size4,.katex .sizing.reset-size3.size4{font-size:1.14285714em}.katex .fontsize-ensurer.reset-size3.size5,.katex .sizing.reset-size3.size5{font-size:1.28571429em}.katex .fontsize-ensurer.reset-size3.size6,.katex .sizing.reset-size3.size6{font-size:1.42857143em}.katex .fontsize-ensurer.reset-size3.size7,.katex .sizing.reset-size3.size7{font-size:1.71428571em}.katex .fontsize-ensurer.reset-size3.size8,.katex .sizing.reset-size3.size8{font-size:2.05714286em}.katex .fontsize-ensurer.reset-size3.size9,.katex .sizing.reset-size3.size9{font-size:2.46857143em}.katex .fontsize-ensurer.reset-size3.size10,.katex .sizing.reset-size3.size10{font-size:2.96285714em}.katex .fontsize-ensurer.reset-size3.size11,.katex .sizing.reset-size3.size11{font-size:3.55428571em}.katex .fontsize-ensurer.reset-size4.size1,.katex .sizing.reset-size4.size1{font-size:.625em}.katex .fontsize-ensurer.reset-size4.size2,.katex .sizing.reset-size4.size2{font-size:.75em}.katex .fontsize-ensurer.reset-size4.size3,.katex .sizing.reset-size4.size3{font-size:.875em}.katex .fontsize-ensurer.reset-size4.size4,.katex .sizing.reset-size4.size4{font-size:1em}.katex .fontsize-ensurer.reset-size4.size5,.katex .sizing.reset-size4.size5{font-size:1.125em}.katex .fontsize-ensurer.reset-size4.size6,.katex .sizing.reset-size4.size6{font-size:1.25em}.katex .fontsize-ensurer.reset-size4.size7,.katex .sizing.reset-size4.size7{font-size:1.5em}.katex .fontsize-ensurer.reset-size4.size8,.katex .sizing.reset-size4.size8{font-size:1.8em}.katex .fontsize-ensurer.reset-size4.size9,.katex .sizing.reset-size4.size9{font-size:2.16em}.katex .fontsize-ensurer.reset-size4.size10,.katex .sizing.reset-size4.size10{font-size:2.5925em}.katex .fontsize-ensurer.reset-size4.size11,.katex .sizing.reset-size4.size11{font-size:3.11em}.katex .fontsize-ensurer.reset-size5.size1,.katex .sizing.reset-size5.size1{font-size:.55555556em}.katex .fontsize-ensurer.reset-size5.size2,.katex .sizing.reset-size5.size2{font-size:.66666667em}.katex .fontsize-ensurer.reset-size5.size3,.katex .sizing.reset-size5.size3{font-size:.77777778em}.katex .fontsize-ensurer.reset-size5.size4,.katex .sizing.reset-size5.size4{font-size:.88888889em}.katex .fontsize-ensurer.reset-size5.size5,.katex .sizing.reset-size5.size5{font-size:1em}.katex .fontsize-ensurer.reset-size5.size6,.katex .sizing.reset-size5.size6{font-size:1.11111111em}.katex .fontsize-ensurer.reset-size5.size7,.katex .sizing.reset-size5.size7{font-size:1.33333333em}.katex .fontsize-ensurer.reset-size5.size8,.katex .sizing.reset-size5.size8{font-size:1.6em}.katex .fontsize-ensurer.reset-size5.size9,.katex .sizing.reset-size5.size9{font-size:1.92em}.katex .fontsize-ensurer.reset-size5.size10,.katex .sizing.reset-size5.size10{font-size:2.30444444em}.katex .fontsize-ensurer.reset-size5.size11,.katex .sizing.reset-size5.size11{font-size:2.76444444em}.katex .fontsize-ensurer.reset-size6.size1,.katex .sizing.reset-size6.size1{font-size:.5em}.katex .fontsize-ensurer.reset-size6.size2,.katex .sizing.reset-size6.size2{font-size:.6em}.katex .fontsize-ensurer.reset-size6.size3,.katex .sizing.reset-size6.size3{font-size:.7em}.katex .fontsize-ensurer.reset-size6.size4,.katex .sizing.reset-size6.size4{font-size:.8em}.katex .fontsize-ensurer.reset-size6.size5,.katex .sizing.reset-size6.size5{font-size:.9em}.katex .fontsize-ensurer.reset-size6.size6,.katex .sizing.reset-size6.size6{font-size:1em}.katex .fontsize-ensurer.reset-size6.size7,.katex .sizing.reset-size6.size7{font-size:1.2em}.katex .fontsize-ensurer.reset-size6.size8,.katex .sizing.reset-size6.size8{font-size:1.44em}.katex .fontsize-ensurer.reset-size6.size9,.katex .sizing.reset-size6.size9{font-size:1.728em}.katex .fontsize-ensurer.reset-size6.size10,.katex .sizing.reset-size6.size10{font-size:2.074em}.katex .fontsize-ensurer.reset-size6.size11,.katex .sizing.reset-size6.size11{font-size:2.488em}.katex .fontsize-ensurer.reset-size7.size1,.katex .sizing.reset-size7.size1{font-size:.41666667em}.katex .fontsize-ensurer.reset-size7.size2,.katex .sizing.reset-size7.size2{font-size:.5em}.katex .fontsize-ensurer.reset-size7.size3,.katex .sizing.reset-size7.size3{font-size:.58333333em}.katex .fontsize-ensurer.reset-size7.size4,.katex .sizing.reset-size7.size4{font-size:.66666667em}.katex .fontsize-ensurer.reset-size7.size5,.katex .sizing.reset-size7.size5{font-size:.75em}.katex .fontsize-ensurer.reset-size7.size6,.katex .sizing.reset-size7.size6{font-size:.83333333em}.katex .fontsize-ensurer.reset-size7.size7,.katex .sizing.reset-size7.size7{font-size:1em}.katex .fontsize-ensurer.reset-size7.size8,.katex .sizing.reset-size7.size8{font-size:1.2em}.katex .fontsize-ensurer.reset-size7.size9,.katex .sizing.reset-size7.size9{font-size:1.44em}.katex .fontsize-ensurer.reset-size7.size10,.katex .sizing.reset-size7.size10{font-size:1.72833333em}.katex .fontsize-ensurer.reset-size7.size11,.katex .sizing.reset-size7.size11{font-size:2.07333333em}.katex .fontsize-ensurer.reset-size8.size1,.katex .sizing.reset-size8.size1{font-size:.34722222em}.katex .fontsize-ensurer.reset-size8.size2,.katex .sizing.reset-size8.size2{font-size:.41666667em}.katex .fontsize-ensurer.reset-size8.size3,.katex .sizing.reset-size8.size3{font-size:.48611111em}.katex .fontsize-ensurer.reset-size8.size4,.katex .sizing.reset-size8.size4{font-size:.55555556em}.katex .fontsize-ensurer.reset-size8.size5,.katex .sizing.reset-size8.size5{font-size:.625em}.katex .fontsize-ensurer.reset-size8.size6,.katex .sizing.reset-size8.size6{font-size:.69444444em}.katex .fontsize-ensurer.reset-size8.size7,.katex .sizing.reset-size8.size7{font-size:.83333333em}.katex .fontsize-ensurer.reset-size8.size8,.katex .sizing.reset-size8.size8{font-size:1em}.katex .fontsize-ensurer.reset-size8.size9,.katex .sizing.reset-size8.size9{font-size:1.2em}.katex .fontsize-ensurer.reset-size8.size10,.katex .sizing.reset-size8.size10{font-size:1.44027778em}.katex .fontsize-ensurer.reset-size8.size11,.katex .sizing.reset-size8.size11{font-size:1.72777778em}.katex .fontsize-ensurer.reset-size9.size1,.katex .sizing.reset-size9.size1{font-size:.28935185em}.katex .fontsize-ensurer.reset-size9.size2,.katex .sizing.reset-size9.size2{font-size:.34722222em}.katex .fontsize-ensurer.reset-size9.size3,.katex .sizing.reset-size9.size3{font-size:.40509259em}.katex .fontsize-ensurer.reset-size9.size4,.katex .sizing.reset-size9.size4{font-size:.46296296em}.katex .fontsize-ensurer.reset-size9.size5,.katex .sizing.reset-size9.size5{font-size:.52083333em}.katex .fontsize-ensurer.reset-size9.size6,.katex .sizing.reset-size9.size6{font-size:.5787037em}.katex .fontsize-ensurer.reset-size9.size7,.katex .sizing.reset-size9.size7{font-size:.69444444em}.katex .fontsize-ensurer.reset-size9.size8,.katex .sizing.reset-size9.size8{font-size:.83333333em}.katex .fontsize-ensurer.reset-size9.size9,.katex .sizing.reset-size9.size9{font-size:1em}.katex .fontsize-ensurer.reset-size9.size10,.katex .sizing.reset-size9.size10{font-size:1.20023148em}.katex .fontsize-ensurer.reset-size9.size11,.katex .sizing.reset-size9.size11{font-size:1.43981481em}.katex .fontsize-ensurer.reset-size10.size1,.katex .sizing.reset-size10.size1{font-size:.24108004em}.katex .fontsize-ensurer.reset-size10.size2,.katex .sizing.reset-size10.size2{font-size:.28929605em}.katex .fontsize-ensurer.reset-size10.size3,.katex .sizing.reset-size10.size3{font-size:.33751205em}.katex .fontsize-ensurer.reset-size10.size4,.katex .sizing.reset-size10.size4{font-size:.38572806em}.katex .fontsize-ensurer.reset-size10.size5,.katex .sizing.reset-size10.size5{font-size:.43394407em}.katex .fontsize-ensurer.reset-size10.size6,.katex .sizing.reset-size10.size6{font-size:.48216008em}.katex .fontsize-ensurer.reset-size10.size7,.katex .sizing.reset-size10.size7{font-size:.57859209em}.katex .fontsize-ensurer.reset-size10.size8,.katex .sizing.reset-size10.size8{font-size:.69431051em}.katex .fontsize-ensurer.reset-size10.size9,.katex .sizing.reset-size10.size9{font-size:.83317261em}.katex .fontsize-ensurer.reset-size10.size10,.katex .sizing.reset-size10.size10{font-size:1em}.katex .fontsize-ensurer.reset-size10.size11,.katex .sizing.reset-size10.size11{font-size:1.19961427em}.katex .fontsize-ensurer.reset-size11.size1,.katex .sizing.reset-size11.size1{font-size:.20096463em}.katex .fontsize-ensurer.reset-size11.size2,.katex .sizing.reset-size11.size2{font-size:.24115756em}.katex .fontsize-ensurer.reset-size11.size3,.katex .sizing.reset-size11.size3{font-size:.28135048em}.katex .fontsize-ensurer.reset-size11.size4,.katex .sizing.reset-size11.size4{font-size:.32154341em}.katex .fontsize-ensurer.reset-size11.size5,.katex .sizing.reset-size11.size5{font-size:.36173633em}.katex .fontsize-ensurer.reset-size11.size6,.katex .sizing.reset-size11.size6{font-size:.40192926em}.katex .fontsize-ensurer.reset-size11.size7,.katex .sizing.reset-size11.size7{font-size:.48231511em}.katex .fontsize-ensurer.reset-size11.size8,.katex .sizing.reset-size11.size8{font-size:.57877814em}.katex .fontsize-ensurer.reset-size11.size9,.katex .sizing.reset-size11.size9{font-size:.69453376em}.katex .fontsize-ensurer.reset-size11.size10,.katex .sizing.reset-size11.size10{font-size:.83360129em}.katex .fontsize-ensurer.reset-size11.size11,.katex .sizing.reset-size11.size11{font-size:1em}.katex .delimsizing.size1{font-family:KaTeX_Size1}.katex .delimsizing.size2{font-family:KaTeX_Size2}.katex .delimsizing.size3{font-family:KaTeX_Size3}.katex .delimsizing.size4{font-family:KaTeX_Size4}.katex .delimsizing.mult .delim-size1>span{font-family:KaTeX_Size1}.katex .delimsizing.mult .delim-size4>span{font-family:KaTeX_Size4}.katex .nulldelimiter{display:inline-block;width:.12em}.katex .delimcenter,.katex .op-symbol{position:relative}.katex .op-symbol.small-op{font-family:KaTeX_Size1}.katex .op-symbol.large-op{font-family:KaTeX_Size2}.katex .accent>.vlist-t,.katex .op-limits>.vlist-t{text-align:center}.katex .accent .accent-body{position:relative}.katex .accent .accent-body:not(.accent-full){width:0}.katex .overlay{display:block}.katex .mtable .vertical-separator{display:inline-block;min-width:1px}.katex .mtable .arraycolsep{display:inline-block}.katex .mtable .col-align-c>.vlist-t{text-align:center}.katex .mtable .col-align-l>.vlist-t{text-align:left}.katex .mtable .col-align-r>.vlist-t{text-align:right}.katex .svg-align{text-align:left}.katex svg{fill:currentColor;stroke:currentColor;fill-rule:nonzero;fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke-miterlimit:4;stroke-dasharray:none;stroke-dashoffset:0;stroke-opacity:1;display:block;height:inherit;position:absolute;width:100%}.katex svg path{stroke:none}.katex img{border-style:none;max-height:none;max-width:none;min-height:0;min-width:0}.katex .stretchy{display:block;overflow:hidden;position:relative;width:100%}.katex .stretchy:after,.katex .stretchy:before{content:""}.katex .hide-tail{overflow:hidden;position:relative;width:100%}.katex .halfarrow-left{left:0;overflow:hidden;position:absolute;width:50.2%}.katex .halfarrow-right{overflow:hidden;position:absolute;right:0;width:50.2%}.katex .brace-left{left:0;overflow:hidden;position:absolute;width:25.1%}.katex .brace-center{left:25%;overflow:hidden;position:absolute;width:50%}.katex .brace-right{overflow:hidden;position:absolute;right:0;width:25.1%}.katex .x-arrow-pad{padding:0 .5em}.katex .cd-arrow-pad{padding:0 .55556em 0 .27778em}.katex .mover,.katex .munder,.katex .x-arrow{text-align:center}.katex .boxpad{padding:0 .3em}.katex .fbox,.katex .fcolorbox{border:.04em solid;box-sizing:border-box}.katex .cancel-pad{padding:0 .2em}.katex .cancel-lap{margin-left:-.2em;margin-right:-.2em}.katex .sout{border-bottom-style:solid;border-bottom-width:.08em}.katex .angl{border-right:.049em solid;border-top:.049em solid;box-sizing:border-box;margin-right:.03889em}.katex .anglpad{padding:0 .03889em}.katex .eqn-num:before{content:"(" counter(katexEqnNo) ")";counter-increment:katexEqnNo}.katex .mml-eqn-num:before{content:"(" counter(mmlEqnNo) ")";counter-increment:mmlEqnNo}.katex .mtr-glue{width:50%}.katex .cd-vert-arrow{display:inline-block;position:relative}.katex .cd-label-left{display:inline-block;position:absolute;right:calc(50% + .3em);text-align:left}.katex .cd-label-right{display:inline-block;left:calc(50% + .3em);position:absolute;text-align:right}.katex-display{display:block;margin:1em 0;text-align:center}.katex-display>.katex{display:block;text-align:center;white-space:nowrap}.katex-display>.katex>.katex-html{display:block;position:relative}.katex-display>.katex>.katex-html>.tag{position:absolute;right:0}.katex-display.leqno>.katex>.katex-html>.tag{left:0;right:auto}.katex-display.fleqn>.katex{padding-left:2em;text-align:left}body{counter-reset:katexEqnNo mmlEqnNo}

/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.katex-error {
	color: var(--vscode-editorError-foreground);
}

/* From extension ms-toolsai.jupyter */
/* These classnames are inherited from bootstrap, but are present in most notebook renderers */

.alert {
    width: auto;
    padding: 1em;
    margin-top: 1em;
    margin-bottom: 1em;
}
.alert > *:last-child {
    margin-bottom: 0;
}
#preview > .alert:last-child {
    /* Prevent this being set to zero by the default notebook stylesheet */
    padding-bottom: 1em;
}

.alert-success {
    /* Note there is no suitable color available, so we just copy "info" */
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-info {
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-warning {
    background-color: var(--theme-warning-background);
    color: var(--theme-warning-foreground);
}
.alert-danger {
    background-color: var(--theme-error-background);
    color: var(--theme-error-foreground);
}

</style>
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.10.2/dist/katex.min.css" integrity="sha384-yFRtMMDnQtDRO8rLpMIKrtPCD5jdktao2TV19YiZYWMDkUR5GQZR/NOVTdquEx1j" crossorigin="anonymous">
<link href="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.css" rel="stylesheet" type="text/css">
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item { list-style-type: none; } .task-list-item-checkbox { margin-left: -20px; vertical-align: middle; }
</style>
        
    </head>
    <body class="vscode-body vscode-light">
        <hr>
<p>title: &quot;PiGx-BSseq: Differential Methylation Report&quot;
author: &quot;BIMSB Bioinformatics Platform&quot;
date: '<code>r format(as.POSIXct(if (&quot;&quot; != Sys.getenv(&quot;SOURCE_DATE_EPOCH&quot;)) { as.numeric(Sys.getenv(&quot;SOURCE_DATE_EPOCH&quot;)) } else { Sys.time() }, origin=&quot;1970-01-01&quot;), &quot;%Y-%m-%d %H:%M:%S&quot;)</code>'
link-citations: yes
output:
html_notebook:
toc : TRUE
depth : 2
toc_float : TRUE
theme : 'lumen'
number_sections : TRUE
code_folding : &quot;hide&quot;
self_contained : TRUE
bibliography : 'report_templates/reports.bib'
params:
scripts_dir:            ''
cpgIsland_bedfile:      ''
refGenes_bedfile:       ''
chrom_seqlengths:       ''
qvalue:                 0.01
difference:             25
webfetch:               ''</p>
<p>sampleids:              ''
treatments:             ''
assembly:               ''
context:                ''
destranded:             ''</p>
<p>control_group:          ''
treatment_group:        ''</p>
<p>methylDiff_file:        ''
methylBase_file:	      ''
methylDiff_results_file: ''</p>
<h2 id="prefix-----------------workdir----------------logo-------------------cores------------------8">prefix:                 ''
workdir:                ''
logo:                   ''
cores:                  8</h2>
<style>
#logo
{
    position: relative;
}
#logo img {
    /*position: relative;*/
    top: 25px;
    /*right: 0px;*/
    left: 50px;
    position: fixed;
    width: 125px;
    }
body
{
    position: absolute;
    top: 150px;
}
</style>
<div id="logo" align="top">
```{r echo=FALSE}
knitr::include_graphics(params$logo)
```
</div>
<pre><code class="language-{r"><code><div>knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.height = 4.5, fig.width = 8, fig.show = &quot;hold&quot;)
knitr::opts_knit$set(progress = FALSE)

## load libraries
library(&quot;methylKit&quot;)
library(&quot;DT&quot;)
library(&quot;genomation&quot;)
library(&quot;rtracklayer&quot;)
library(&quot;ggplot2&quot;)
library(&quot;ggrepel&quot;)
require(&quot;matrixStats&quot;)
require(&quot;data.table&quot;)
data.table::setDTthreads(params$cores)

</div></code></code></pre>
<h1 id="description">Description</h1>
<p>PiGx BSseq performs differential methylation testing using <em>methylKit's</em> ,<code>calculateDiffMeth()</code> function and produces this report. The report includes tables and figures summarizing similarities and differences between comparison groups as specified in the settings file.</p>
<p>This report was generated with PiGx BSseq version @VERSION@.</p>
<pre><code class="language-{r"><code><div>
hyper.col &lt;- &quot;magenta&quot;
hypo.col &lt;- &quot;darkolivegreen4&quot;
nTop &lt;- 5000

cpgIsland_bedfile       &lt;- params$cpgIsland_bedfile
refGenes_bedfile        &lt;- params$refGenes_bedfile
webfetch                &lt;- tolower(params$webfetch) %in% c(&quot;true&quot;, &quot;yes&quot;)
chrom_seqlengths_file   &lt;- params$chrom_seqlengths

methylDiff_file &lt;- params$methylDiff_file
methylBase_file &lt;- params$methylBase_file
methylDiff_results_file &lt;- params$methylDiff_results_file

qvalue &lt;- as.numeric(params$qvalue)
difference &lt;- as.numeric(params$difference)

scripts_dir  &lt;- params$scripts_dir

sampleids     &lt;- strsplit(params$sampleids, &quot;,&quot;, fixed = FALSE, perl = FALSE, useBytes = FALSE)[[1]]
# split all treatment values, could be numeric or not
treatmentsStr  &lt;- strsplit(params$treatments, &quot;,&quot;, fixed = FALSE, perl = FALSE, useBytes = FALSE)[[1]]
# convert into named numeric vector
treatments &lt;- as.numeric( as.factor( treatmentsStr ))
names(treatments) &lt;- treatmentsStr

assembly  &lt;- params$assembly
context &lt;- params$context
destranded &lt;- ifelse(tolower(params$destranded) %in% c(&quot;true&quot;,&quot;yes&quot;),TRUE,FALSE)


control_group_str &lt;- params$control_group
control_group &lt;- strsplit(params$control_group, &quot;,&quot;, fixed = FALSE, perl = FALSE, useBytes = FALSE)[[1]]
treatment_group_str &lt;- params$treatment_group
treatment_group &lt;- strsplit(params$treatment_group, &quot;,&quot;, fixed = FALSE, perl = FALSE, useBytes = FALSE)[[1]]

if (length(control_group) &gt; 0) {
  # reorganize samples into treatment and control groups
  case_group &lt;- ifelse(treatmentsStr %in% treatment_group,&quot;Treatment&quot;,&quot;Control&quot;)
} else {
  # convert into named numeric vector
  case_group &lt;- rep(length(treatmentsStr),&quot;Treatment&quot; )
}

prefix &lt;-  params$prefix
workdir &lt;- params$workdir
cores &lt;-   params$cores

## Some sections are only run if respective output is available
ExtractDiffMeth   &lt;- TRUE
AnnotateDiffMeth  &lt;- TRUE
ExportDMR &lt;- TRUE
AnnotateDMR &lt;- TRUE

if(! (all(file.exists(cpgIsland_bedfile,refGenes_bedfile)) | webfetch )) {
  AnnotateDiffMeth &lt;- FALSE
}

#create a folder to save high quality images generated by the report script
imagesDir &lt;- file.path(workdir, paste0(prefix, '_images'))
if(!dir.exists(imagesDir)) {
  dir.create(path = imagesDir)
}

inputParameterDesc &lt;- c('Control sample groups',
                        'Treatment sample groups',
                        'Qvalue cutoff',
                        'Minimal Methylation Difference',
                        'United Samples Tabix file',
                        'DiffMeth Tabix file',
                        'Genome Assembly',
                        'Prefix for output files',
                        'Working directory'
                     )
inputParameterValues &lt;- c(
    control_group_str,
    treatment_group_str,
    qvalue,
    difference,
    basename(methylDiff_file),
    basename(methylBase_file),
    assembly,
    prefix,
    workdir
)

inputSettings &lt;- data.frame(parameters = inputParameterDesc,
                            values = inputParameterValues,
                            stringsAsFactors = FALSE)
DT::datatable(data = inputSettings,
              extensions = 'FixedColumns',
              options = list(fixedColumns = TRUE,
                         scrollX = TRUE,
                         pageLength = 9,
                         dom = 't'))

</div></code></code></pre>
<h1 id="quality-control">Quality Control</h1>
<h2 id="sample-clustering--top-r-ntop-most-variable-sites-">Sample Clustering ( top <code>r nTop</code> most variable sites )</h2>
<p>The first quality control should be to check wether the samples cluster in their respective groups, or wether there are some known or unknown batch effects that need to be considered for the following differential methylation analysis. This plot shows the result of a principle components analysis (PCA) based on the <code>r nTop</code> most variable cytosine sites.</p>
<pre><code class="language-{r"><code><div>
#' Get Most Variable CpGs Percent Methylation
#'
#' This function extracts the top most variable
#'
#' @param .object   can be a methylBase, methylBaseDB or
#'   methylation matrix extracted by percMethylation()
#' @param file      path to file storing methylation matrix extracted by percMethylation()
#' @param nTop      number of top most variable sites
#' @param chunk.size  how big are the chunks to be processed (only relevant for methylBaseDB )
#'
#' @return methylation matrix ordered by variability
getMostVariableCpGs = function(.object=NULL, file=NULL, nTop=5000, chunk.size = 1e6){
    require(matrixStats)
    require(data.table)
    data.table::setDTthreads(8)

    if ( !is.null(.object) ) {
        if (class(.object) %in% c(&quot;methylBase&quot;,&quot;methylBaseDB&quot;) ) {
        meth.mat.dt &lt;- as.data.table(percMethylation(.object,chunk.size = chunk.size))
        } else {
            meth.mat.dt = copy(as.data.table(.object))
        }
    } else if ( !is.null(file) &amp; file.exists(file) )  {
        meth.mat.dt &lt;- data.table::fread(file)
    }

    meth.mat.dt[, variance := rowVars(as.matrix(.SD))]
    ## fastest ordering method
    ## https://stackoverflow.com/questions/13685295/sort-a-data-table-fast-by-ascending-descending-order
    meth.mat.dt[, id := seq_len(nrow(meth.mat.dt))]
    setorder(meth.mat.dt, -variance)
    meth.mat.dt = meth.mat.dt[1:min(nrow(meth.mat.dt),nTop),]
    meth.mat.dt[, `:=`(variance=NULL, id=NULL)]
    return(meth.mat.dt)

}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
# Read input files
methylBaseDB &lt;- methylKit:::readMethylBaseDB(dbpath = methylBase_file,
                                    dbtype = &quot;tabix&quot;,
                                    sample.ids = sampleids,
                                    assembly = assembly,
                                    context = context,
                                    resolution = &quot;base&quot;,
                                    treatment = treatments,
                                    destranded = destranded)


topVar &lt;- getMostVariableCpGs(methylBaseDB, nTop = nTop)
pca &lt;- stats::prcomp(t(topVar), center = TRUE)
pcaSummary &lt;- summary(pca)
df &lt;- cbind(as.data.frame(pca$x), &quot;Group&quot; = treatmentsStr, &quot;Case&quot; = case_group)
df$SampleID &lt;- row.names(df)

</div></code></code></pre>
<pre><code class="language-{r"><code><div>for ( batch in c(&quot;Group&quot;,&quot;Case&quot;)) {

p &lt;- ggplot(df, aes(x = PC1, y = PC2)) +
    geom_point(aes_string(color = batch)) +
    geom_label_repel(aes(label = SampleID), size = 3) +
    labs(x = paste0('PC1 (',round(pcaSummary$importance[2, 'PC1'] * 100, 1),'%)'),
         y = paste0('PC2 (',round(pcaSummary$importance[2, 'PC2'] * 100, 1),'%)')) +
    ggtitle(label = paste(&quot;PCA plot based on&quot;,nrow(topVar),&quot;most variable&quot;, context,&quot; sites&quot;)) +
    theme_bw()

print(p)

ggsave(plot = p, filename = file.path(imagesDir,paste0(prefix,&quot;.pca.png&quot;)),
       width = 7,height = 6)
}

</div></code></code></pre>
<h1 id="differential-methylation-analysis">Differential Methylation Analysis</h1>
<h2 id="calling-differentially-methylated-cytosines">Calling Differentially Methylated Cytosines</h2>
<p>For each treatment vector specified in the  settings file, the ratio of methylation between the control (i.e. the first entry), and the treatment (i.e. the second entry) is calculated across the genome. Logistic regression is then applied to model the log-odds probability of observing this ratio, by chance, in any given location (p-value) as well as the probability of observing it <em>somewhere</em> within the genome (q-value). The resulting probabilities are then tabulated below.</p>
<p>After q-value calculation, differentially methylated bases are extracted based on q-value and percent methylation difference cutoffs. Here we select bases that have q-value &lt; <code>r qvalue</code> and percent methylation difference larger than <code>r difference</code>%. Furthermore, we calculate hyper-methylated or hypo-methylated bases.</p>
<p>Overdispersion occurs when there is more variability in the data than assumed by the distribution and is here included in the differentially methylatation calculation.</p>
<p>For more details about <code>calculateDiffMeth()</code> and <code>getMethylDiff()</code> functions see <a href="https://bioconductor.org/packages/release/bioc/html/methylKit.html">methylKit</a> and for details about
the logistic regression and overdispersion see <a href="https://www.sciencedirect.com/science/article/pii/S0168165617315936">Wreczycka, et al (2017)</a>.</p>
<p>The full table of methylKit calculateDiffMeth() results can be found at:</p>
<ul>
<li><strong>methylKit calculateDiffMeth results file</strong>:
<code>r methylDiff_results_file</code></li>
</ul>
<pre><code class="language-{r"><code><div>

if (!setequal(control_group,treatment_group)) {
  message(&quot;Remapping Treatments Descriptions into treatment and control groups.&quot;)
  # reorganize samples into treatment and control groups
  treatments_cmp &lt;- ifelse(treatmentsStr %in% treatment_group,1,0)

  message(&quot;Control group (0): &quot;, paste0(control_group,collapse = &quot;, &quot;))
  message(&quot;Treatment group (1): &quot;, paste0(treatment_group,collapse = &quot;, &quot;))
} else {
  message(&quot;Remapping Treatment Description to number.&quot;)
  # convert into named numeric vector
  treatments_cmp &lt;- as.numeric( as.factor( treatmentsStr ))
  names(treatments_cmp) &lt;- treatmentsStr

  message(paste(sort(unique(treatmentsStr)),
                sort(unique(treatments_cmp)),
                sep = &quot;: &quot;,collapse = &quot;\n&quot;))

}


methylDiffDB &lt;- methylKit:::readMethylDiffDB(dbpath = methylDiff_file,
                                    dbtype = &quot;tabix&quot;,
                                    sample.ids = sampleids,
                                    assembly = assembly,
                                    context = context,
                                    resolution = &quot;base&quot;,
                                    treatment = treatments_cmp,
                                    destranded = destranded)

# Get differentially methylated bases based on cutoffs
methylDiff.obj.all &lt;- getMethylDiff(.Object = methylDiffDB,
									difference=difference,
									type=&quot;all&quot;,
									qvalue = qvalue,
									save.db = FALSE)

# Check if there are some differentially methylated cytosines
if (! nrow(methylDiff.obj.all)&gt;1 ) {
	cat(&quot;\n**There are no differentially methylated cytosines.**&quot;)
	ExtractDiffMeth &lt;- FALSE
	AnnotateDiffMeth &lt;- FALSE
} else {
    cat(&quot;\n**We found &quot;,nrow(methylDiff.obj.all),&quot; cytosines to be differentially methylated .**&quot;)
  }

</div></code></code></pre>
<pre><code class="language-{r"><code><div>cat('### Export Differentially Methylated Cytosines\n\n')

cat('We export differentially-methylated CpG sites (DMC) to a *BED* file; it can be loaded into a genome browser such as [IGV](http://software.broadinstitute.org/software/igv/) or [UCSC](https://genome.ucsc.edu/) to allow for further analysis, annotation and visualisation.\n\n')

cat('- **DmC BED file**:\n',file.path(workdir,paste0(prefix,&quot;.dmc.bed&quot;)),'\n')

</div></code></code></pre>
<pre><code class="language-{r"><code><div>
# Get hyper-methylated
methylDiff.obj.hyper &lt;- getMethylDiff(.Object = methylDiff.obj.all,
									  difference = difference,
									  type = &quot;hyper&quot;,
									  qvalue = qvalue)
# Get hypo-methylated
methylDiff.obj.hypo &lt;- getMethylDiff(.Object = methylDiff.obj.all,
									 difference = difference,
									 type = &quot;hypo&quot;,
									 qvalue=qvalue)

</div></code></code></pre>
<pre><code class="language-{r"><code><div>
#-------------------------------------------------------------------
#' Export windows to a BED file
#'
#' The windows are color coded based on their score (methylation or
#' differential methylation value). If score is only positive
#' (e.g. methylation value) score range will be mapped to [0,1] range,
#' if score contains negative values (e.g. methylation difference) colors are
#' centered first to [-1,1] and then mapped to [0,1] range.
#'
#' @param windows \code{\link[GenomicRanges]{GRanges}} object with information about
#' differentially methylated regions
#' @param filename name of the output data
#' @param colramp color scale to be used in the BED display
#' defaults to gray,green, darkgreen scale.
#' You may use divergent color scale as well, e.g.
#' colorRamp(c(&quot;darkolivegreen4&quot;,&quot;grey&quot;,&quot;magenta&quot;)).
#' @param scoreCol name of the column to use as score,
#' defaults to &quot;meth.diff&quot;
#' @param normalize_score should the score be normalized to range [0,1] ?
#' @param track_name name of the track display
#' @param description additional description
#' @param skipTrackLine do not write trackline
#' @param coord_only generate a plain BED file with only ranges,
#' for direct use in Great or Enrichr webservers
#'
#' @return A BED files with the differentially methylated regions
#' which can be visualized in the UCSC browser
#'
#' @seealso \code{\link{methylKit::methSeg2bed}}
#'
#' @export
#' @docType methods
#' @rdname meth2bed
meth2bed&lt;-function(windows,filename,
                   scoreCol = &quot;meth.diff&quot;,
                   normalize_score = FALSE,
                   track_name='meth windows',
                   description = 'meth windows',
                   colramp=colorRamp(c(&quot;gray&quot;,&quot;green&quot;, &quot;darkgreen&quot;)),
                   skipTrackLine = FALSE,
                   coord_only = FALSE
){
    require(rtracklayer)

    range01 &lt;- function(x){ (x-min(x)+1)/(max(x)-min(x)+1) }
    rangePlusMinus &lt;- function(x) { ((x + 1 ) / 2 ) }


    if(class(windows)!=&quot;GRanges&quot;){
        windows=as(windows, &quot;GRanges&quot;)
    }

    strand(windows)=&quot;*&quot;
    if(! scoreCol %in% names(mcols(windows)) ) {
        stop(&quot;Could not find column &quot;,scoreCol,&quot; in given windows.&quot;)
    } else {
        score(windows)=mcols(windows)[[scoreCol]]
        score_temp &lt;-  score(windows)
        if( any(score_temp &lt; 0 ) ) {
            if(any(score_temp &gt; 1)) {score_temp = score_temp/100 }
           score_temp=rangePlusMinus(score_temp)
        } else {
           score_temp=range01(score_temp)
        }
    }

    trackLine &lt;- sprintf(&quot;track name='%s' description='%s' itemRgb=On&quot;,
                         track_name,
                         description)

    ## case if only one line is exported
    if(is.null(colramp) | length(windows)==1){
        trackLine &lt;- gsub(pattern = &quot;itemRgb=On&quot;,replacement = &quot;&quot;,x = trackLine)
    } else {
        ramp &lt;- colramp
        mcols(windows)$itemRgb= rgb(ramp(score_temp), maxColorValue = 255)
    }

    if(normalize_score) {
        score(windows) &lt;- score_temp
    }

    if(is.null(filename)) {
        return(as.data.frame(windows))
    }

    if(coord_only){
        export.bed(granges(windows),filename)
    } else if(skipTrackLine){
        export.bed(windows,filename)
    } else {
        export.bed(windows,filename,
                   trackLine=as(trackLine, &quot;BasicTrackLine&quot;))
    }
}

# Export differentially methylated cytosines to a bed file
bedFile &lt;- file.path(workdir,paste0(prefix,&quot;.dmc.bed&quot;))
meth2bed(windows = methylDiff.obj.all,
         track_name = sprintf(&quot;DMC (%s)&quot;,
                              gsub(&quot;_destranded|methylDackel&quot;,&quot;&quot;,prefix)),
         description = sprintf(&quot;Differential methylation between %s and %s &quot;,
                               treatment_group_str,control_group_str),
         scoreCol = &quot;meth.diff&quot;,
         colramp=colorRamp(c(&quot;gray&quot;,&quot;green&quot;, &quot;darkgreen&quot;)),
         filename = bedFile)


</div></code></code></pre>
<pre><code class="language-{r"><code><div>cat('## Finding Regions of Differential Methylation\n\n')

cat('We use the `methSeg()` function to perform an unsupervised segmentation of the methylation differences calculated from `calculateDiffMeth()`. Then we classify hyper and hypo methylated segments based on the methylation difference threshold and join neighbouring segments with the same class. For details about the segmentation see [Wreczycka, et al (2017)](https://www.sciencedirect.com/science/article/pii/S0168165617315936). ')

cat('To aggregate the region level statistics we use Stoufferâ€™s weighted Z-method as done by  [methcp](https://github.com/boyinggong/methcp), for details see [Gong and Purdom  (2018)](https://www.biorxiv.org/content/10.1101/265116v2). ')


</div></code></code></pre>
<pre><code class="language-{r"><code><div>source(file.path(scripts_dir, &quot;find_dmrs.R&quot;))

pdf(file.path(workdir,paste0(prefix,&quot;.find_dmr.pdf&quot;)))
dmrs &lt;- find_DMR(methylDiffDB,cores = cores)
invisible(dev.off())

if(! length(unique(dmrs$seg.group)) &gt; 1 ) {
  cat(
    &quot;\n**The segments belong only to a single class,\n&quot;,
    &quot;This could indicate the unsupervised clustering of the &quot;,
    &quot;segments was not successful or the difference in methylation &quot;,
    &quot;is not high enough.**&quot;
  )
	ExportDMR &lt;- FALSE
	AnnotateDMR &lt;- FALSE
} else {
    cat(&quot;\n**The segmentation was successful, out of &quot;,length(dmrs),&quot; regions &quot;,
    &quot;we found &quot;,sum(dmrs$seg.group == &quot;hyper&quot;) ,&quot;to be hyper-methylated and &quot;,
    sum(dmrs$seg.group == &quot;hypo&quot;), &quot; to be hypo-methylated.**&quot;)
  }
</div></code></code></pre>
<pre><code class="language-{r"><code><div>cat('### Export Differentially Methylated Regions\n\n')

cat('We export the regions of differentially-methylated cytosines (DMRs) to a *BED* file; it can be loaded into a genome browser such as [IGV](http://software.broadinstitute.org/software/igv/) or [UCSC](https://genome.ucsc.edu/) to allow for further analysis, annotation and visualisation.\n\n')

cat('If you intend to perform online enrichment analysis using webservices like [GREAT](http://great.stanford.edu/public/html/index.php) or [Enrichr](http://amp.pharm.mssm.edu/Enrichr/) please use the \'coordOnly\' BED file.\n\n')

dmSegmentFile &lt;- file.path(workdir,paste0(prefix,&quot;.diffmethSegments.bed&quot;))
dmSegmentTable &lt;- file.path(workdir,paste0(prefix,&quot;.diffmethSegments.tsv&quot;))
dmrCoordOnlyFile &lt;- file.path(workdir,paste0(prefix,&quot;.coordOnly.bed&quot;))
dmrFile &lt;- file.path(workdir,paste0(prefix,&quot;.dmr.bed&quot;))

cat('- **DMR BED file**:\n',dmrFile,'\n')
cat('- **DMR BED file (coordinates only) **:\n',dmrCoordOnlyFile,'\n')
cat('- **Genome Wide Differential Methylation BED file**:\n',dmSegmentFile,'\n')
cat('- **Genome Wide Differential Methylation TSV file**:\n',dmSegmentTable,'\n')

</div></code></code></pre>
<pre><code class="language-{r"><code><div>

meth2bed(
  dmrs,
  scoreCol = &quot;seg.mean&quot;,
  filename = dmSegmentFile,
  colramp = colorRamp(c(hypo.col, &quot;grey&quot;, hyper.col)),
  track_name = sprintf(&quot;DM (%s)&quot;,
                       gsub(pattern = sprintf(&quot;_%s_.*&quot;, context),
                            &quot;&quot;,
                            prefix)),
  description = &quot;Genome Wide Differential Methylation&quot;

)

dmrs.df &lt;- as.data.frame(dmrs)
colnames(dmrs.df) &lt;- gsub(&quot;seg.mean&quot;,&quot;meth.diff&quot;,colnames(dmrs.df))
dmrs.df$ID &lt;- NULL
dmrs.df &lt;- dmrs.df[order(dmrs.df$meth.diff,decreasing = TRUE),]
write.table(
  x = as.data.frame(dmrs.df),
  file = dmSegmentTable,
  sep = &quot;\t&quot;,
  row.names = FALSE,
  quote = FALSE
)
rm(dmrs.df)

meth2bed(
  dmrs[dmrs$seg.group %in% c(&quot;hyper&quot;, &quot;hypo&quot;)],
  scoreCol = &quot;seg.mean&quot;,
  filename = dmrFile,
  colramp = colorRamp(c(hypo.col, &quot;grey&quot;, hyper.col)),
  track_name = sprintf(&quot;DMR (%s)&quot;,
                       gsub(pattern = sprintf(&quot;_%s_.*&quot;, context),
                            &quot;&quot;,
                            prefix)),
  description = paste(
    &quot;Differentially methylated regions with&quot;,
    &quot;methylation difference &gt; &quot;,
    difference
  )
)

meth2bed(
  windows = dmrs[dmrs$seg.group %in% c(&quot;hyper&quot;, &quot;hypo&quot;)],
  scoreCol = &quot;seg.mean&quot;,
  filename = dmrCoordOnlyFile,
  coord_only = TRUE
)
</div></code></code></pre>
<pre><code class="language-{r"><code><div>cat('# Annotation of Differential Methylation\n\n')

cat('## Annotation of Differentially Methylated Cytosines\n\n')

cat('Once differentially-methylated CpG sites (DMC) have been found, the sites are classified into hyper- and hypo-methylated sites, based on two cutoffs: the minimal  methylation difference and the qvalue cutoff as shown in the parameter table. Hypermethylation indicates increasingly-methylated sites, while hypomethylation indicates lowered methylation when performing the comparison of treatment - control.\n\n')
</div></code></code></pre>
<pre><code class="language-{r"><code><div>chrom_seqlengths &lt;- read.delim(chrom_seqlengths_file,
                               header = FALSE, stringsAsFactors = FALSE)

myseqinfo &lt;- Seqinfo(seqnames = chrom_seqlengths[,1],
                     seqlengths = chrom_seqlengths[,2],
                     genome = assembly)
myseqinfo = keepStandardChromosomes(myseqinfo)
myseqinfo &lt;- sortSeqlevels(myseqinfo)

</div></code></code></pre>
<pre><code class="language-{r"><code><div># Convert a methylDiff object to a GRanges object
GRanges.diffmeth    = as(methylDiff.obj.all, &quot;GRanges&quot;)
GRanges.diffmeth &lt;- sortSeqlevels(GRanges.diffmeth)
# subset to all available chromosomes
seqlevels(myseqinfo) &lt;- seqlevels(GRanges.diffmeth)
seqinfo(GRanges.diffmeth) &lt;- myseqinfo

GRanges.diffmeth.hyper  = as(methylDiff.obj.hyper, &quot;GRanges&quot;)
GRanges.diffmeth.hyper &lt;- sortSeqlevels(GRanges.diffmeth.hyper)

GRanges.diffmeth.hypo   = as(methylDiff.obj.hypo, &quot;GRanges&quot;)
GRanges.diffmeth.hypo &lt;- sortSeqlevels(GRanges.diffmeth.hypo)
# GRanges.diffmeth.nonsig = as(methylDiff.obj.nonsig, &quot;GRanges&quot;)
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if(length(GRanges.diffmeth.hypo)&gt;1 &amp; length(GRanges.diffmeth.hyper)&gt;1){

cat('### Overview of Hyper- and Hypo-Methylated CpGs Over the Genome\n\n')

cat('This chromosome ideogram represents differential methylation, where only CpGs fulfilling the cutoffs of q-value &lt; ', qvalue ,' and methylation difference of at least ', difference,'% are shown.\n' )

}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if(length(GRanges.diffmeth.hypo)&gt;1 &amp; length(GRanges.diffmeth.hyper)&gt;1){

source(file.path(scripts_dir, &quot;ideoDMC.R&quot;))
ideoDMC_hyper_hypo(hyper = GRanges.diffmeth.hyper,
                   hypo = GRanges.diffmeth.hypo,
                   chrom.length = seqlengths( sortSeqlevels(myseqinfo) ),
                   circos = FALSE,
                   title = &quot;Genome Wide Differential Methylation&quot;,
                   hyper.col = hyper.col, hypo.col = hypo.col) + theme_classic()

}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>cat('#### Fraction of Differentially Methylated Cytosines per Chromosome\n\n')

cat('These stacked barplots visualize the percentages of hyper and hypomethylated DMCs out of all covered CpGs for each chromosome.\n\n')
</div></code></code></pre>
<pre><code class="language-{r"><code><div># Get number of differentially methylated cytosines per chromosome
stats.perChr &lt;- diffMethPerChr(methylDiffDB,plot = FALSE)$diffMeth.per.chr

stats.df &lt;- reshape2::melt(stats.perChr,
               measure.vars = c(&quot;number.of.hypermethylated&quot;,
                                &quot;number.of.hypomethylated&quot;),
               id.vars = &quot;chr&quot;,
               value.name = &quot;count&quot;)
levels(stats.df$chr) &lt;- unique(stats.df$chr)
stats.df$perc &lt;- reshape2::melt(stats.perChr,
               measure.vars = c(&quot;percentage.of.hypermethylated&quot;,
                                &quot;percentage.of.hypomethylated&quot;),
               id.vars = &quot;chr&quot;,
               value.name = &quot;percentage&quot;)[,&quot;percentage&quot;]

stats.df$variable &lt;- factor(gsub(&quot;number.of.&quot;,&quot;&quot;,stats.df$variable),
                            levels = c(&quot;hypomethylated&quot;,&quot;hypermethylated&quot;))
</div></code></code></pre>
<pre><code class="language-{r"><code><div>p &lt;- ggplot(stats.df, aes(x = chr, y = perc, fill = variable)) +
  scale_fill_manual(values = c(hypo.col,hyper.col)) +
  geom_bar(stat = &quot;identity&quot;,position = &quot;stack&quot;, color = &quot;black&quot;) +
  geom_text_repel(aes(label = count), position = position_stack(vjust = 0.5),
            hjust = 0.1,size = 5) +
  ggtitle(label = &quot;Differentially methylated Cytosines per Chromosome&quot;) +
  labs(y = &quot;Percentage&quot;, x = &quot;Chromosome&quot;) +
  coord_flip() +
  theme_bw() +
  theme(legend.position = &quot;bottom&quot;, legend.title = element_blank())

print(p)
</div></code></code></pre>
<pre><code class="language-{r"><code><div>cat('These stacked barplots visualize the percentages of hyper and hypomethylated DMCs out of all detected differential CpGs for each chromosome.\n\n')
</div></code></code></pre>
<pre><code class="language-{r"><code><div># Get number of differentially methylated cytosines per chromosome
stats.DMC.perChr &lt;- diffMethPerChr(methylDiff.obj.all,plot = FALSE)$diffMeth.per.chr

stats.DMC.df &lt;- reshape2::melt(stats.DMC.perChr,
               measure.vars = c(&quot;number.of.hypermethylated&quot;,
                                &quot;number.of.hypomethylated&quot;),
               id.vars = &quot;chr&quot;,
               value.name = &quot;count&quot;)
levels(stats.DMC.df$chr) &lt;- unique(stats.DMC.df$chr)
stats.DMC.df$perc &lt;- reshape2::melt(stats.DMC.perChr,
               measure.vars = c(&quot;percentage.of.hypermethylated&quot;,
                                &quot;percentage.of.hypomethylated&quot;),
               id.vars = &quot;chr&quot;,
               value.name = &quot;percentage&quot;)[,&quot;percentage&quot;]

stats.DMC.df$variable &lt;- factor(gsub(&quot;number.of.&quot;,&quot;&quot;,stats.DMC.df$variable),
                            levels = c(&quot;hypomethylated&quot;,&quot;hypermethylated&quot;))

p &lt;- ggplot(stats.DMC.df, aes(x = chr, y = perc, fill = variable)) +
  scale_fill_manual(values = c(hypo.col,hyper.col)) +
  geom_bar(stat = &quot;identity&quot;,position = &quot;stack&quot;, color = &quot;black&quot;) +
  geom_text_repel(aes(label = count), position = position_stack(vjust = 0.5),
            hjust = 0.1,size = 5) +
  ggtitle(label = &quot;Differentially methylated Cytosines per Chromosome&quot;) +
  labs(y = &quot;Percentage&quot;, x = &quot;Chromosome&quot;) +
  coord_flip() +
  theme_bw() +
  theme(legend.position = &quot;bottom&quot;, legend.title = element_blank())

print(p)
</div></code></code></pre>
<pre><code class="language-{r"><code><div>source(file.path(scripts_dir, &quot;fetch_procedures.R&quot;))
fetched.refgenes &lt;- lookupBedFile(type = &quot;knownGene&quot;,
                               filename = refGenes_bedfile,
                               assembly = assembly,
                               webfetch = webfetch )
fetch_refgen_success &lt;- !is.null(fetched.refgenes)
fetched.cpgi &lt;- lookupBedFile(type = &quot;cpgIslandExt&quot;,
                              filename =  cpgIsland_bedfile,
                              assembly = assembly,
                              webfetch = webfetch )

fetch_cpgi_success &lt;- !is.null(fetched.cpgi)

</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.diffmeth)!=0){
  ## now we parse the gene features
  refgenes.grl &lt;- readTranscriptFeatures(fetched.refgenes)

  annot.gene &lt;- annotateWithGeneParts(target = GRanges.diffmeth,
                                         feature = refgenes.grl,
                                         intersect.chr = TRUE)

  if( length(GRanges.diffmeth.hypo)!=0 &amp; length(GRanges.diffmeth.hyper)!=0){

  annot.gene.hyper &lt;- annotateWithGeneParts(target = GRanges.diffmeth.hyper,
                                         feature = refgenes.grl,
                                         intersect.chr = TRUE)

  annot.gene.hypo &lt;- annotateWithGeneParts(target = GRanges.diffmeth.hypo,
                                         feature = refgenes.grl,
                                         intersect.chr = TRUE)
  }


  cat('### Differentially Methylated Cytosines per Genomic Feature\n\n')

  cat('The distribution of differentially methylated cytosines over the genomic features is a good start to check hypothesis that could indicate possible effects on the gene expression level.\n\n')

}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.diffmeth)!=0){

  cat('#### Piechart\n\n')
  # # plot the target overlap for each
  genomation::plotTargetAnnotation(annot.gene,
                       main=&quot;Differentially methylated Cytosines\nper Genomic Feature&quot;)

  cat('\n\n')
}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if( (fetch_refgen_success) &amp; length(GRanges.diffmeth)!=0){

  # construct dataframe for ggplot
  df &lt;- reshape2::melt(getTargetAnnotationStats(annot.gene))
  df$variable &lt;- factor(names(getTargetAnnotationStats(annot.gene)),
                        levels = rev(names(getTargetAnnotationStats(annot.gene))))
  df$count &lt;- annot.gene@num.annotation

  cols &lt;- getColors(length(getTargetAnnotationStats(annot.gene)))



  p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
    scale_fill_manual(values = rev(cols)) +
    geom_bar(stat = &quot;identity&quot; ,color = &quot;black&quot;, position = position_dodge()) +
    geom_text(aes(y = -10, label = count), hjust = 0,size = 5) +
    ggtitle(label = &quot;Differentially methylated Cytosines per Genomic Feature&quot;) +
    labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
    coord_flip() +
    theme_bw() +
    guides(fill = guide_legend(reverse = TRUE)) +
    theme(legend.position = &quot;bottom&quot;, legend.title = element_blank())

  cat('#### Barchart\n\n')

  print(p)

  cat('\n\n')


}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
fill_strip &lt;- function(facet_ggplot, fills, plot=FALSE) {
  # extract gtable
  g &lt;- ggplot_gtable(ggplot_build(facet_ggplot))
  # figure out facet strips
  strip_both &lt;- which(grepl('strip-', g$layout$name))
  #

  if(! length(fills) == length(strip_both)) {
    stop(&quot;Length of colors (&quot;,length(fills),&quot;)&quot;,
         &quot;does not match with number of strips (&quot;,
         length(strip_both),&quot;).&quot;)
  }

  k &lt;- 1
  for (i in strip_both) {
  j &lt;- which(grepl('rect', g$grobs[[i]]$grobs[[1]]$childrenOrder))
  g$grobs[[i]]$grobs[[1]]$children[[j]]$gp$fill &lt;- fills[k]
  k &lt;- k+1
  }

  if(plot) {
    grid.newpage(recording = FALSE)
    grid.draw(g)
  }

  invisible(return(g))

}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>  if( length(GRanges.diffmeth.hypo)!=0 &amp; length(GRanges.diffmeth.hyper)!=0){


    df &lt;- reshape2::melt(cbind(getTargetAnnotationStats(annot.gene.hyper),
                               getTargetAnnotationStats(annot.gene.hypo)))
    names(df) &lt;- c(&quot;variable&quot;,&quot;type&quot;,&quot;value&quot;)
    df$variable &lt;- factor(df$variable,
                        levels = rev(names(getTargetAnnotationStats(annot.gene))))
    df$type &lt;- factor(ifelse(df$type == 1,&quot;hypermethylated&quot;,&quot;hypomethylated&quot;),
                      levels = c(&quot;hypomethylated&quot;,&quot;hypermethylated&quot;))
    df$count &lt;- c(annot.gene.hyper@num.annotation,
                      annot.gene.hypo@num.annotation)

    p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
      scale_fill_manual(values = rev(cols)) +
      geom_bar(mapping = aes(group = type), stat = &quot;identity&quot; ,color = &quot;black&quot;,  position = position_dodge()) +
      geom_text(mapping = aes(group = type, color = type , y = -10, label = count), size = 5,
                  position = position_dodge(width= 1)) +
      scale_color_manual(values = c(hypo.col,hyper.col)) +
      ggtitle(label = &quot;Differentially methylated Cytosines per Genomic Feature&quot;) +
      labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
      coord_flip() +
      theme_bw() +
      guides(fill = guide_legend(reverse = TRUE)) +
      theme(legend.position = &quot;bottom&quot;, legend.title = element_blank(),
            legend.text = element_text(colour=c(rep(times = length(cols),&quot;black&quot;),
                                                c(hypo.col,hyper.col))))

    cat('#### Barchart (Split per DMC Type)\n\n')

    p &lt;- p + facet_wrap( ~ type,nrow = 2)

    print(p)

    # png(filename = file.path(imagesDir,paste0(prefix,&quot;DMCperGenomicFeat.type.png&quot;)),
    #        height = 4.5, width = 8)
    # g &lt;- fill_strip(p, fills = rev(c(hypo.col,hyper.col)),plot = TRUE)
    # dev.off()
    # # ggsave(plot = g, filename = file.path(imagesDir,paste0(prefix,&quot;DMCperGenomicFeat.type.png&quot;)),
    #        # width = 7,height = 6)
    #
    # cat('![](',file.path(imagesDir,paste0(prefix,&quot;DMCperGenomicFeat.type.png&quot;)),')')

    cat('\n\n')
  }
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_cpgi_success) &amp; length(GRanges.diffmeth)!=0){

  # read the shores and flanking regions and name the flanks as shores
  # and CpG islands as CpGi
    cpg.obj=readFeatureFlank(fetched.cpgi,
                           feature.flank.name=c(&quot;CpGi&quot;,&quot;shores&quot;))
  #
  # convert methylDiff object to GRanges and annotate
  diffCpGann=annotateWithFeatureFlank(target = GRanges.diffmeth,
                                      feature = cpg.obj$CpGi,
                                      flank = cpg.obj$shores,
                                      feature.name=&quot;CpGi&quot;,flank.name=&quot;shores&quot;)

  if( length(GRanges.diffmeth.hypo)!=0 &amp; length(GRanges.diffmeth.hyper)!=0){

      diffCpGann.hypo=annotateWithFeatureFlank(target = GRanges.diffmeth.hypo,
                                          feature = cpg.obj$CpGi,
                                          flank = cpg.obj$shores,
                                          feature.name=&quot;CpGi&quot;,flank.name=&quot;shores&quot;)

      diffCpGann.hyper=annotateWithFeatureFlank(target = GRanges.diffmeth.hyper,
                                          feature = cpg.obj$CpGi,
                                          flank = cpg.obj$shores,
                                          feature.name=&quot;CpGi&quot;,flank.name=&quot;shores&quot;)
  }


  cat('### Differentially Methylated Cytosines over CpG Islands\n\n')

    cat(&quot;CpG islands are genomic features of 200 to 2000 base pairs that have an increased CG base pair content and show abundant CpG sequences. A large fraction (~70%) of human promoters of a gene contain a CpG island ([Deaton, A. M., &amp; Bird, A. (2011)](http://genesdev.cshlp.org/content/25/10/1010.short)). Silencing of genes is often achieved by methylation in CpG Islands, so observation of hypomethylation in CpG Islands could indicate effects on the expression of a gene.\n\n&quot;)


}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_cpgi_success) &amp; length(GRanges.diffmeth)!=0){

  cat('#### Piechart\n\n')

  # plot the target overlap for each
  genomation::plotTargetAnnotation(diffCpGann,
                       col=c(&quot;green&quot;,&quot;gray&quot;,&quot;white&quot;),
                       main=&quot;Differentially methylated Cytosines over CpG Islands\n&quot;)

  cat('\n\n')
}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if( (fetch_cpgi_success) &amp; length(GRanges.diffmeth)!=0){

  # construct dataframe for ggplot
  df &lt;- reshape2::melt(getTargetAnnotationStats(diffCpGann))
  df$variable &lt;- factor(names(getTargetAnnotationStats(diffCpGann)),
                        levels = rev(names(getTargetAnnotationStats(diffCpGann))))
  df$count &lt;- diffCpGann@num.annotation

  p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
  scale_fill_manual(values = c(&quot;white&quot;,&quot;gray&quot;,&quot;green&quot;)) +
  geom_bar(stat = &quot;identity&quot; ,color = &quot;black&quot;, position = position_dodge()) +
    geom_text(aes(y = -10, label = count), hjust = 0,size = 5) +
  coord_flip() +
  labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
  ggtitle(label = &quot;Differentially methylated Cytosines per CpG Island Feature&quot;) +
  theme_bw() +
  guides(fill = guide_legend(reverse = TRUE)) +
  theme(legend.position = &quot;bottom&quot;,
        legend.title = element_blank(),
        # axis.title.y = element_blank(), axis.ticks.y = element_blank(),
        # axis.text.y = element_blank()
        )

  cat('#### Barchart\n\n')

  print(p)

  cat('\n\n')
}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if( length(GRanges.diffmeth.hypo)!=0 &amp; length(GRanges.diffmeth.hyper)!=0){

  # construct dataframe for ggplot
  df &lt;- reshape2::melt(cbind(getTargetAnnotationStats(diffCpGann.hyper),
                             getTargetAnnotationStats(diffCpGann.hypo)))
  names(df) &lt;- c(&quot;variable&quot;,&quot;type&quot;,&quot;value&quot;)
  df$variable &lt;- factor(df$variable,
                        levels = rev(names(getTargetAnnotationStats(diffCpGann))))
  df$type &lt;- factor(ifelse(df$type == 1,&quot;hypermethylated&quot;,&quot;hypomethylated&quot;),
                    levels = c(&quot;hypomethylated&quot;,&quot;hypermethylated&quot;))
  df$count &lt;- c(diffCpGann.hyper@num.annotation,diffCpGann.hypo@num.annotation)


  p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
  scale_fill_manual(values = c(&quot;white&quot;,&quot;gray&quot;,&quot;green&quot;)) +
  geom_bar(mapping = aes(group = type), stat = &quot;identity&quot; ,color = &quot;black&quot;,  position = position_dodge()) +
  geom_text(mapping = aes(group = type, color = type , y = -10, label = count), size = 5,
            position = position_dodge(width= 1)) +
  scale_color_manual(values = c(hypo.col,hyper.col)) +
  coord_flip() +
  labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
  ggtitle(label = &quot;Differentially methylated Cytosines per CpG Island Feature&quot;) +
  theme_bw() +
  guides(fill = guide_legend(reverse = TRUE),color = FALSE) +
  theme(legend.position = &quot;bottom&quot;,
        legend.title = element_blank(),
        # axis.title.y = element_blank(), axis.ticks.y = element_blank(),
        # axis.text.y = element_blank()
        )

  cat('#### Barchart (Split per DMC Type)\n\n')

  p &lt;- p + facet_wrap(~type, nrow = 2)

  print(p)

  cat('\n\n')

}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.diffmeth)!=0){

  cat('### Distribution of differential methylation around TSS\n\n')

  cat(&quot;This histogram of differential CpG methylation show how far the different types of differentially methylated cytosines are distributed around the closest transcription start site of the nearest genes.\n\n&quot;)

}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.diffmeth)!=0)
  {
  # Get distance to nearest TSS and gene id from AnnotationByGeneParts
  # target.row is the row number in diffmeth.gr
  assoTSS = getAssociationWithTSS(annot.gene)

  assoTSS$dmc.type &lt;- &quot;hypermethylated&quot;
  assoTSS$dmc.type[ GRanges.diffmeth[assoTSS$target.row]$meth.diff &lt; 0] &lt;- &quot;hypomethylated&quot;

  p &lt;- ggplot(assoTSS, aes(x = dist.to.feature, fill = dmc.type)) +
    geom_histogram(color = &quot;black&quot;, position = position_dodge()) +
    scale_fill_manual(values = c(hyper.col,hypo.col)) +
    labs(x=&quot;Base Pairs&quot;, y=&quot;Frequency&quot; ) +
    ggtitle(&quot;Distance to the nearest TSS&quot;) +
    theme_bw() +
    theme(legend.position = &quot;bottom&quot;,
          legend.title = element_blank(),
          )

  print(p)

  }

</div></code></code></pre>
<pre><code class="language-{r"><code><div>cat('## Annotation of Differentially Methylated Regions\n\n')

cat('Once regions of consecutive differential methylation (DMR) have been found for the comparison of treatment - control, those regions are classified into hyper- and hypo-methylated or unchanged regions, based on the minimal methylation difference as shown in the parameter table. Hypermethylation indicates increasingly-methylated sites, while hypomethylation indicates lowered methylation when performing the comparison of treatment - control.\n\n')
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
# DMRs are already GRanges
GRanges.dmr &lt;- sortSeqlevels(dmrs[dmrs$seg.group!=&quot;none&quot;])
names(mcols(GRanges.dmr)) &lt;- gsub(&quot;seg.mean&quot;,&quot;meth.diff&quot;, names(mcols(GRanges.dmr)))

# subset to all available chromosomes
seqlevels(GRanges.dmr, pruning.mode = &quot;coarse&quot;) &lt;- seqlevels(myseqinfo)
seqinfo(GRanges.dmr) &lt;- myseqinfo

# Subset DMRs
GRanges.dmr.hyper  = GRanges.dmr[GRanges.dmr$seg.group == &quot;hyper&quot;]
GRanges.dmr.hyper &lt;- sortSeqlevels(GRanges.dmr.hyper)

GRanges.dmr.hypo   = GRanges.dmr[GRanges.dmr$seg.group == &quot;hypo&quot;]
GRanges.dmr.hypo &lt;- sortSeqlevels(GRanges.dmr.hypo)
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if(length(GRanges.dmr.hypo)&gt;1 &amp; length(GRanges.dmr.hyper)&gt;1){

cat('### Overview of Hyper- and Hypo-Methylated DMRs Over the Genome\n\n')

cat('This chromosome ideogram represents differential methylation, where only regions of Cytosines fulfilling the cutoff of methylation difference of at least ', difference,'% are shown.\n' )

}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if(length(GRanges.dmr.hypo)&gt;1 &amp; length(GRanges.dmr.hyper)&gt;1){

source(file.path(scripts_dir, &quot;ideoDMC.R&quot;))
ideoDMC_hyper_hypo(hyper = GRanges.dmr.hyper,
                   hypo = GRanges.dmr.hypo,
                   chrom.length = seqlengths( sortSeqlevels(myseqinfo) ),
                   circos = FALSE,
                   title = &quot;Genome Wide Differential Methylation (DMRs)&quot;,
                   hyper.col = hyper.col, hypo.col = hypo.col) + theme_classic()

}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.dmr)!=0){
  ## now we parse the gene features
  refgenes.grl &lt;- readTranscriptFeatures(fetched.refgenes)

  annot.dmr.gene &lt;- annotateWithGeneParts(target = GRanges.dmr,
                                         feature = refgenes.grl,
                                         intersect.chr = TRUE)

  if( length(GRanges.dmr.hypo)!=0 &amp; length(GRanges.dmr.hyper)!=0){

  annot.dmr.gene.hyper &lt;- annotateWithGeneParts(target = GRanges.dmr.hyper,
                                         feature = refgenes.grl,
                                         intersect.chr = TRUE)

  annot.dmr.gene.hypo &lt;- annotateWithGeneParts(target = GRanges.dmr.hypo,
                                         feature = refgenes.grl,
                                         intersect.chr = TRUE)
  }


  cat('### Differentially Methylated Regions per Genomic Feature\n\n')

  cat('The distribution of regions of differentially methylated cytosines over the genomic features should match the annotation based on DMCs and provides an additional confirmation for the stated hypothesis.\n\n')

}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.dmr)!=0){

  cat('#### Piechart\n\n')
  # # plot the target overlap for each
  genomation::plotTargetAnnotation(annot.dmr.gene,
                       main=&quot;Differentially methylated Regions\nper Genomic Feature&quot;)

  cat('\n\n')
}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if( (fetch_refgen_success) &amp; length(GRanges.dmr)!=0){

  # construct dataframe for ggplot
  df &lt;- reshape2::melt(getTargetAnnotationStats(annot.dmr.gene))
  df$variable &lt;- factor(names(getTargetAnnotationStats(annot.dmr.gene)),
                        levels = rev(names(getTargetAnnotationStats(annot.dmr.gene))))
  df$count &lt;- annot.dmr.gene@num.annotation

  cols &lt;- getColors(length(getTargetAnnotationStats(annot.dmr.gene)))



  p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
    scale_fill_manual(values = rev(cols)) +
    geom_bar(stat = &quot;identity&quot; ,color = &quot;black&quot;, position = position_dodge()) +
    geom_text(aes(y = -10, label = count), hjust = 0,size = 5) +
    ggtitle(label = &quot;Differentially methylated Regions per Genomic Feature&quot;) +
    labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
    coord_flip() +
    theme_bw() +
    guides(fill = guide_legend(reverse = TRUE)) +
    theme(legend.position = &quot;bottom&quot;, legend.title = element_blank())

  cat('#### Barchart\n\n')

  print(p)

  cat('\n\n')


}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>  if( length(GRanges.dmr.hypo)!=0 &amp; length(GRanges.dmr.hyper)!=0){


    df &lt;- reshape2::melt(cbind(getTargetAnnotationStats(annot.dmr.gene.hyper),
                               getTargetAnnotationStats(annot.dmr.gene.hypo)))
    names(df) &lt;- c(&quot;variable&quot;,&quot;type&quot;,&quot;value&quot;)
    df$variable &lt;- factor(df$variable,
                        levels = rev(names(getTargetAnnotationStats(annot.dmr.gene))))
    df$type &lt;- factor(ifelse(df$type == 1,&quot;hypermethylated&quot;,&quot;hypomethylated&quot;),
                      levels = c(&quot;hypomethylated&quot;,&quot;hypermethylated&quot;))
    df$count &lt;- c(annot.dmr.gene.hyper@num.annotation,
                      annot.dmr.gene.hypo@num.annotation)

    p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
      scale_fill_manual(values = rev(cols)) +
      geom_bar(mapping = aes(group = type), stat = &quot;identity&quot; ,color = &quot;black&quot;,  position = position_dodge()) +
      geom_text(mapping = aes(group = type, color = type , y = -10, label = count), size = 5,
                  position = position_dodge(width= 1)) +
      scale_color_manual(values = c(hypo.col,hyper.col)) +
      ggtitle(label = &quot;Differentially methylated Regions per Genomic Feature&quot;) +
      labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
      coord_flip() +
      theme_bw() +
      guides(fill = guide_legend(reverse = TRUE)) +
      theme(legend.position = &quot;bottom&quot;, legend.title = element_blank(),
            legend.text = element_text(colour=c(rep(times = length(cols),&quot;black&quot;),
                                                c(hypo.col,hyper.col))))

    cat('#### Barchart (Split per DMR Type)\n\n')

    p &lt;- p + facet_wrap( ~ type,nrow = 2)

    print(p)

    # png(filename = file.path(imagesDir,paste0(prefix,&quot;DMCperGenomicFeat.type.png&quot;)),
    #        height = 4.5, width = 8)
    # g &lt;- fill_strip(p, fills = rev(c(hypo.col,hyper.col)),plot = TRUE)
    # dev.off()
    # # ggsave(plot = g, filename = file.path(imagesDir,paste0(prefix,&quot;DMCperGenomicFeat.type.png&quot;)),
    #        # width = 7,height = 6)
    #
    # cat('![](',file.path(imagesDir,paste0(prefix,&quot;DMCperGenomicFeat.type.png&quot;)),')')

    cat('\n\n')
  }
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_cpgi_success) &amp; length(GRanges.dmr)!=0){

  # read the shores and flanking regions and name the flanks as shores
  # and CpG islands as CpGi
  cpg.obj=readFeatureFlank(fetched.cpgi,
                         feature.flank.name=c(&quot;CpGi&quot;,&quot;shores&quot;))

  # convert methylDiff object to GRanges and annotate
  diffCpGann.dmr=annotateWithFeatureFlank(target = GRanges.dmr,
                                      feature = cpg.obj$CpGi,
                                      flank = cpg.obj$shores,
                                      feature.name=&quot;CpGi&quot;,flank.name=&quot;shores&quot;)

  if( length(GRanges.dmr.hypo)!=0 &amp; length(GRanges.dmr.hyper)!=0){

      diffCpGann.dmr.hypo=annotateWithFeatureFlank(target = GRanges.dmr.hypo,
                                          feature = cpg.obj$CpGi,
                                          flank = cpg.obj$shores,
                                          feature.name=&quot;CpGi&quot;,flank.name=&quot;shores&quot;)

      diffCpGann.dmr.hyper=annotateWithFeatureFlank(target = GRanges.dmr.hyper,
                                          feature = cpg.obj$CpGi,
                                          flank = cpg.obj$shores,
                                          feature.name=&quot;CpGi&quot;,flank.name=&quot;shores&quot;)
  }


  cat('### Differentially Methylated Regions over CpG Islands\n\n')

    cat(&quot;CpG islands are genomic features of 200 to 2000 base pairs that have an increased CG base pair content and show abundant CpG sequences. A large fraction (~70%) of human promoters of a gene contain a CpG island ([Deaton, A. M., &amp; Bird, A. (2011)](http://genesdev.cshlp.org/content/25/10/1010.short)). Silencing of genes is often achieved by methylation in CpG Islands, so observation of hypomethylation in CpG Islands could indicate effects on the expression of a gene.\n\n&quot;)


}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_cpgi_success) &amp; length(GRanges.dmr)!=0){

  cat('#### Piechart\n\n')

  # plot the target overlap for each
  genomation::plotTargetAnnotation(diffCpGann.dmr,
                       col=c(&quot;green&quot;,&quot;gray&quot;,&quot;white&quot;),
                       main=&quot;Differentially methylated Regions over CpG Islands\n&quot;)

  cat('\n\n')
}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if( (fetch_cpgi_success) &amp; length(GRanges.dmr)!=0){

  # construct dataframe for ggplot
  df &lt;- reshape2::melt(getTargetAnnotationStats(diffCpGann.dmr))
  df$variable &lt;- factor(names(getTargetAnnotationStats(diffCpGann.dmr)),
                        levels = rev(names(getTargetAnnotationStats(diffCpGann.dmr))))
  df$count &lt;- diffCpGann.dmr@num.annotation

  p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
  scale_fill_manual(values = c(&quot;white&quot;,&quot;gray&quot;,&quot;green&quot;)) +
  geom_bar(stat = &quot;identity&quot; ,color = &quot;black&quot;, position = position_dodge()) +
    geom_text(aes(y = -10, label = count), hjust = 0,size = 5) +
  coord_flip() +
  labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
  ggtitle(label = &quot;Differentially methylated Regions per CpG Island Feature&quot;) +
  theme_bw() +
  guides(fill = guide_legend(reverse = TRUE)) +
  theme(legend.position = &quot;bottom&quot;,
        legend.title = element_blank(),
        # axis.title.y = element_blank(), axis.ticks.y = element_blank(),
        # axis.text.y = element_blank()
        )

  cat('#### Barchart\n\n')

  print(p)

  cat('\n\n')
}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>
if( length(GRanges.dmr.hypo)!=0 &amp; length(GRanges.dmr.hyper)!=0){

  # construct dataframe for ggplot
  df &lt;- reshape2::melt(cbind(getTargetAnnotationStats(diffCpGann.dmr.hyper),
                             getTargetAnnotationStats(diffCpGann.dmr.hypo)))
  names(df) &lt;- c(&quot;variable&quot;,&quot;type&quot;,&quot;value&quot;)
  df$variable &lt;- factor(df$variable,
                        levels = rev(names(getTargetAnnotationStats(diffCpGann.dmr))))
  df$type &lt;- factor(ifelse(df$type == 1,&quot;hypermethylated&quot;,&quot;hypomethylated&quot;),
                    levels = c(&quot;hypomethylated&quot;,&quot;hypermethylated&quot;))
  df$count &lt;- c(diffCpGann.dmr.hyper@num.annotation,diffCpGann.dmr.hypo@num.annotation)


  p &lt;- ggplot(df, aes(x = variable, y = value, fill = variable)) +
  scale_fill_manual(values = c(&quot;white&quot;,&quot;gray&quot;,&quot;green&quot;)) +
  geom_bar(mapping = aes(group = type), stat = &quot;identity&quot; ,color = &quot;black&quot;,  position = position_dodge()) +
  geom_text(mapping = aes(group = type, color = type , y = -10, label = count), size = 5,
            position = position_dodge(width= 1)) +
  scale_color_manual(values = c(hypo.col,hyper.col)) +
  coord_flip() +
  labs(y = &quot;Percentage&quot;, x = &quot;Feature&quot;) +
  ggtitle(label = &quot;Differentially methylated Regions per CpG Island Feature&quot;) +
  theme_bw() +
  guides(fill = guide_legend(reverse = TRUE),color = FALSE) +
  theme(legend.position = &quot;bottom&quot;,
        legend.title = element_blank(),
        # axis.title.y = element_blank(), axis.ticks.y = element_blank(),
        # axis.text.y = element_blank()
        )

  cat('#### Barchart (Split per DMR Type)\n\n')

  p &lt;- p + facet_wrap(~type, nrow = 2)

  print(p)

  cat('\n\n')

}

</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.dmr)!=0){

  cat('### Distribution of differential methylation around TSS\n\n')

  cat(&quot;This histogram of differential CpG methylation show how far the different types of differentially methylated regions are distributed around the closest transcription start site of the nearest genes.\n\n&quot;)

}
</div></code></code></pre>
<pre><code class="language-{r"><code><div>if( (fetch_refgen_success) &amp; length(GRanges.dmr)!=0)
  {
  # Get distance to nearest TSS and gene id from AnnotationByGeneParts
  # target.row is the row number in diffmeth.gr
  assoTSS = getAssociationWithTSS(annot.dmr.gene)

  assoTSS$dmc.type &lt;- &quot;hypermethylated&quot;
  assoTSS$dmc.type[ GRanges.dmr[assoTSS$target.row]$meth.diff &lt; 0] &lt;- &quot;hypomethylated&quot;

  p &lt;- ggplot(assoTSS, aes(x = dist.to.feature, fill = dmc.type)) +
    geom_histogram(color = &quot;black&quot;, position = position_dodge()) +
    scale_fill_manual(values = c(hyper.col,hypo.col)) +
    labs(x=&quot;Base Pairs&quot;, y=&quot;Frequency&quot; ) +
    ggtitle(&quot;Distance to the nearest TSS&quot;) +
    theme_bw() +
    theme(legend.position = &quot;bottom&quot;,
          legend.title = element_blank(),
          )

  print(p)

  }

</div></code></code></pre>
<h2 id="session-information">Session Information</h2>
<pre><code class="language-{r"><code><div>sessionInfo()
</div></code></code></pre>

        <script async src="https://cdn.jsdelivr.net/npm/katex-copytex@latest/dist/katex-copytex.min.js"></script>
        
    </body>
    </html>